### PR TITLE
Support complex `addUtilities()` configs

### PR DESCRIPTION
**Pull request by philipp-spiess**
This PR adds support for complex `addUtilities()` configuration objects that use child combinators and other features.

For example, in v3 it was possible to add a utility that changes the behavior of all children of the utility class node by doing something like this:

```ts
addUtilities({
  '.red-children > *': {
    color: 'red',
  },
});
```

This is a pattern that was used by first-party plugins like `@tailwindcss/aspect-ratio` but that we never made working in v4, since it requires parsing the selector and properly extracting all utility candidates.

While working on the codemod that can transform `@layer utilities` scoped declarations like the above, we found out a pretty neat heuristics on how to migrate these cases. We're basically finding all class selectors and replace them with `&`. Then we create a nested CSS structure like this:

```css
.red-children {
  & > * {
    color: red;
  }
}
```

Due to first party support for nesting, this works as expected in v4.

## Test Plan

We added unit tests to ensure the rewriting works in some edge cases. Furthermore we added an integration test running the `@tailwindcss/aspect-ratio` plugin. We've also installed the tarballs in the Remix example from the [playgrounds](https://github.com/philipp-spiess/tailwindcss-playgrounds) and ensure we can use the `@tailwindcss/aspect-ratio` plugin just like we could in v3:
 
Screenshot 2024-11-18 at 13 44 52

### CODE REVIEW

**Review by RobinMalfait — APPROVED** [review]

**Review by thecrypticace — APPROVED** [review]

File: packages/tailwindcss/src/compat/selector-parser.test.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          "nodes": [
+            {
+              "kind": "value",
+              "value": "n+1",
+            },
+          ],
+          "value": ":nth-child",
+        },
+      ]
+    `)
```
Comment (thecrypticace) [inline packages/tailwindcss/src/compat/selector-parser.test.ts:72]:
```suggestion
    expect(parse(':nth-child(n+1)')).toEqual([
      {
        kind: 'function',
        value: ':nth-child',
        nodes: [
          {
            kind: 'value',
            value: 'n+1',
          },
        ],
      },
    ])
```

**Review by thecrypticace — CHANGES_REQUESTED** [review]
a small thing in the tests but otherwise seems good

File: packages/tailwindcss/src/compat/selector-parser.ts [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+        let contents = input.slice(start, end)
+        let node = contents.trim() === ',' ? separator(contents) : combinator(contents)
+        if (parent) {
+          parent.nodes.push(node)
+        } else {
+          ast.push(node)
+        }
+
+        break
+      }
```
[the remark this answers is not in the document]
Comment (philipp-spiess) [reply packages/tailwindcss/src/compat/selector-parser.ts:228]: Ah, thanks, missed the plus sign in a list 😬  Fixed!

**Review by thecrypticace — APPROVED** [review]

**Comment by davidmatter** [comment]
@philipp-spiess Am I right to assume that in v4, we need to migrate from

```js
{
  '@media (min-width: 768px)': {
    .article-display-m {
      font-size: 60px;
    }
  }
}
```
 to:
 ```js
{
  .article-display-m {
    '@media (min-width: 768px)': {
      font-size: 60px;
    }
  }
}
```

**Comment by philipp-spiess** [comment]
@davidmatter Huh this is... interesting! Yes if you can migrate to that that would be easiest. Is this a plugin doing or how did you discover this?

**Comment by davidmatter** [comment]
We're quite a large company with a large CSS-based design system with hundreds of (for example) typography definitions. Now, to get proper intellisense in our code editors and to integrate better with our tailwind projects, we've come up with a custom plugin that parses CSS and adds the definitions to tailwind using `addUtilities`.

e.g.
```
  const root = parse(css)
  const jss = objectify(root)
  addUtilities(jss)
```

It works quite well actually in v3 :) In a perfect world, these definitions would of course be extracted to their own components to support the official tailwind way of doing things etc. but that would be A LOT of work.

**Comment by davidmatter** [comment]
For parsing and `objectify`, we use https://github.com/postcss/postcss-js

**Comment by philipp-spiess** [comment]
I see, so you're not expecting this to generate the right CSS anyways and only use this to extend the type hints for intellisense? I think it should be possible to write a quick function that inverts the object properties in this case?

**Comment by davidmatter** [comment]
@philipp-spiess Thanks for taking the time. Well, we actually use it to generate prunable CSS. TBH I think it's the best pattern to work with legacy CSS code because of 1) Intellisense and 2) Resulting CSS can be pruned by tailwind and runs through the official plugin system.

Inverting the objects is alright for us. I suspect that we're not the only ones using this pattern, though. Might be advisable to mark this as a breaking change to save everyone's time :)

**Comment by davidmatter** [comment]
BTW: Good to see an Austrian engineer being part of tailwind! Keep it up
